components/launcher-protocol/src/types.rs

```diff
@@ -0,0 +1,371 @@
+// Copyright (c) 2019 Chef Software Inc. and/or applicable contributors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::{
+    error::{Error, Result},
+    generated,
+};
+use bytes::BytesMut;
+use prost::{self, Message};
+use std::{collections::HashMap, fmt};
+
+pub trait MessageStatic {
+    const MESSAGE_ID: &'static str;
+}
+
+pub trait FromProto<T: prost::Message + Default>: Sized {
+    fn from_proto(value: T) -> Result<Self>;
+}
+
+pub trait LauncherMessage: MessageStatic + Clone + Default + fmt::Debug {
+    fn to_bytes(&self) -> Result<Vec<u8>>;
+    fn from_bytes(bytes: &[u8]) -> Result<Self>;
+}
```

Comment:
Here's an alternate approach that doesn't need macros:

We can consolidate the `MessageStatic`, and `FromProto` traits into `LauncherMessage`, and by adding an associated type to hold the Prost-generated struct, we get:
```rust
pub trait LauncherMessage
where
    Self: Clone + fmt::Debug,
{
    type Generated: prost::Message + Default + From<Self>;
    const MESSAGE_ID: &'static str;

    fn from_proto(value: Self::Generated) -> Result<Self>;

    fn from_bytes(bytes: &[u8]) -> Result<Self> {
        let decoded = Self::Generated::decode(bytes)?;
        Self::from_proto(decoded)
    }

    fn to_bytes(&self) -> Result<Vec<u8>> {
        let envelope = Self::Generated::from(self.clone());
        let mut buf = bytes::BytesMut::with_capacity(envelope.encoded_len());
        envelope.encode(&mut buf)?;
        Ok(buf.to_vec())
    }
}
```

The `impl` changes follow from that, but [here's a full diff](https://github.com/habitat-sh/habitat/files/2818106/associated-type-refactor.diff.txt) if you want to go that way.

Reply:
This is exactly the solution I was trying to find from the beginning.  Thank you so much for this.